### ISSUE TITLE
PhysicState.get_contact_collider_velocity_at_pos is unusuable in 3D

### ISSUE THREAD

**Issue by kubecz3k** [godotengine/godot#9537]
**Operating system or device - Godot version:**
Linux, Godot 2.1.4 (but I assume this behavior is probably much much older)
update: it's present in 7fc3c51169957f0907a13dd07d4004496e6ebd56 (3.0) as well

**Issue description:**
Recently I was looking at platformer 2d demo, and I saw very neat feature of PhysicState that allows to easily program following after movement floor parts like horizontal platforms in couple lines: (https://github.com/godotengine/godot-demo-projects/blob/2.1/2d/platformer/player.gd#L221). 

```
if (found_floor):
	floor_h_velocity = s.get_contact_collider_velocity_at_pos(floor_index).x
	lv.x += floor_h_velocity
```
![floor_velocity_2d](https://user-images.githubusercontent.com/6129594/27956828-ca354990-631b-11e7-82d4-3015dab24102.gif)

Immediately wanted to apply this to my 3d project where player movement is realized in quite similar fashion (also rigidbody), but it turns out it's not working so well in 3D.
This probably is somewhat know issue to @reduz since I looked at Platformer 3D demo, and very similar code is there... but it's turned off (commented): https://github.com/godotengine/godot-demo-projects/blob/2.1/3d/platformer/player.gd#L169
![floor_velocity_3d](https://user-images.githubusercontent.com/6129594/27956837-d6aab872-631b-11e7-8636-35a9ebd2d125.gif)

Would be cool if we could have something similar in 3d, since at current moment this method is returning big forces even if terrain is perfectly static. 

**UPDATE 1:** For sample and more info check https://github.com/godotengine/godot/issues/9537#issuecomment-332870576

**Steps to reproduce:**
Download Platformer demo 3d and uncomment line 171 of Player script.

**Link to minimal example project:**
**Edit1:** For sample and more info check https://github.com/godotengine/godot/issues/9537#issuecomment-332870576

**Comment by kubecz3k** [comment]
@reduz Maybe you could take a look at this at the time of porting 3dPlatformer demo to 3.0?

**Comment by reduz** [comment]
It should happen in 3D too, it simply doesn't work due to how the 3D platformer demo is made. Will be fixing this today.

**Comment by kubecz3k** [comment]
@reduz is there any reason in particular why new platformer demo is using `fixed_process` instead of `integrate_forces`??

**Comment by kubecz3k** [comment]
In any case, I ported rigid platformer demo to 3.0, and tested it on 7fc3c51169957f0907a13dd07d4004496e6ebd56 unfortunately the results are still the same

**Comment by kubecz3k** [comment]
@AndreaCatania Just wondering if there are chances that this `get_contact_collider_velocity_at_pos` method will work as expected with Bullet. Don't get me wrong, there are ways to workaround this issue, but the simplicity of how moving platforms can be implemented with this method is very nice.

**Comment by AndreaCatania** [comment]
I tested this method `get_contact_collider_velocity_at_pos` time ago when I implemented it, and it looked to me that it is working correctly. Did you test it or it's just a question?
If you tested it and it doesn't work please submit an example project where this function doesn't work so I can fix it.

**Comment by kubecz3k** [comment]
@AndreaCatania it was just a question, but I will test it for sure and let you know :)

**Comment by kubecz3k** [comment]
I checked this issue once again, I can tell for sure `contact_collider_velocity_at_position` is not working well in 3D (or at least it's usage is different than in 2D). I created very minimal reproduction project in 2d which I'm using as a reference on how I think this function should be used. Recreated the same project in 3D (it practically use the same script). Godot physic engine in 3D `contact_collider_velocity_at_position` is returning `0,0,0`. Also tested Bullet branch, and well it's hard for me to tell if this method is working properly on bullet because of other problems. 

Recorded very short 40s vid that is presenting all above: https://youtu.be/jYK5xC_a9fI
Also here is sample project: [FloorVelocityTest.zip](https://github.com/godotengine/godot/files/1341397/FloorVelocityTest.zip)

Reproduction Steps:
1. Open `2DVelocityAtPointTest.tscn`  and run it. Observe proper (I think) behavior.
2. Open `3DVelocityAtPointTest.tscn` and run it. Observe wrong (I think) behavior.
3. (additional) Check together on Godot internal physic engine and on Bullet.

**Comment by kubecz3k** [comment]
@AndreaCatania you might want to check demo/vid from the post above. I'm curious, if you will be able to reproduce how bullet behaves on my machine.

**Comment by reduz** [comment]
The delayed motion is normal. This needs to be fixed in KinematicBody, so
it only actually transforms when it gets the callback from physics server.

**Comment by kubecz3k** [comment]
@reduz what delayed motion?

**Comment by AndreaCatania** [comment]
I'm sure this problem is also on windows because you are using "_integrate_forces"  function. I'll check this problem when the PR of "physics_process" and "get_body_direct_state" will be accepted

**Comment by kubecz3k** [comment]
@AndreaCatania good to know! Will give it a shot then as well :)

**Comment by AndreaCatania** [comment]
I just tested it with https://github.com/godotengine/godot/pull/10013/commits/21535bb7e16ff956dbd6b6bbee98947f684935c9 on windows, and it works.

Here the gif: https://gfycat.com/gifs/detail/ValuableShowyHoopoe

Please let me know if you still have the problem on linux.

**Comment by kubecz3k** [comment]
**_Bullet related post_**
Works very nicely on Linux as well, thank you! I added rotation to the mix. Not sure if this should work the way how I imagine, but it would be very easy to implement rotational moving platforms in such case :) 
vid: https://www.youtube.com/watch?v=cJ3NXgrbg_w&feature=youtu.be
reproduction project: [v2FloorVelocityAndRotationTest.zip](https://github.com/godotengine/godot/files/1384575/v2FloorVelocityAndRotationTest.zip)

**Comment by kubecz3k** [comment]
@AndreaCatania I was trying to play with your recent commits that are getting rotation into account (like we were talking on irc). I haven't found a way to make it usable and I'm not sure but I think there is still something fishy on the bullet site. On this gif I'm visualizing current readings from `state.get_contact_collider_velocity_at_position`: 
![rotationatpoint](https://user-images.githubusercontent.com/6129594/32107589-60c534ea-bb2f-11e7-804e-04de5005013d.gif)

It can be seen the forces are rotating together with the platform, but I think they are rotating in a wrong direction, I imagine this vector should more or less always point in the same direction as that smaller vector from this second gif (since this is more or less direction where that given point will be in next step):
![rotationatpoint2](https://user-images.githubusercontent.com/6129594/32107596-6699cf02-bb2f-11e7-9a4f-862eb30a6b5f.gif)

Here is a sample project which I think is better to use when investigating this:
[FloorRotationAtPointTest.zip](https://github.com/godotengine/godot/files/1422264/FloorRotationAtPointTest.zip)
Also sorry for delay.

**Comment by kubecz3k** [comment]
Also what can be seen on those gifs is that more the more body is near the edge the more its accelerating, which is perfectly good since, points on the edge are traveling greater distance than the point in the middle.

**Comment by kubecz3k** [comment]
Another gif, maybe it will help somehow. It can be seen on it that something is wrong, point floor direction should be constant because body is always touching a point that will move in the same direction as previous one:
![rotationatpoint3](https://user-images.githubusercontent.com/6129594/32109256-7d483054-bb34-11e7-9917-28b52ba79158.gif)

it was made by commenting line 17 in Hero.gd
or if you prefer sample is here: [FloorRotationAtPointTest-no-movement.zip](https://github.com/godotengine/godot/files/1422367/FloorRotationAtPointTest-no-movement.zip)

**Comment by kubecz3k** [comment]
Did some more test. It seems that `get_velocity_at_pos` in some way don't take rotation of the body into account. After compensating for floor body rotation with:
```
var floorRotY = state.get_contact_collider_object(0).rotation.y;
floorVelocity = state.get_contact_collider_velocity_at_position(0);
var fix4FloorVelocity = floorVelocity.rotated(Vector3(0,1,0), floorRotY);
```
It's starting to work for rotation, but in such case movement it broken: 
https://youtu.be/jSlRwITh0SY

project: 
[FloorRotationAtPointTest-compensate-rotation.zip](https://github.com/godotengine/godot/files/1424097/FloorRotationAtPointTest-compensate-rotation.zip)

**Comment by kubecz3k** [comment]
Awesome, its working very well now! :) 
![platforms_fixed](https://user-images.githubusercontent.com/6129594/32135957-6ceaa380-bc07-11e7-8952-812cc7de6991.gif)

This issue can be closed when bullet will be merged.

**Comment by kubecz3k** [comment]
solved by #12641 :smile:
